### PR TITLE
Upload diagnostics to GCS periodically

### PR DESCRIPTION
**Pull request by robputt**
Pushes miners config each time the diagnostics task is run  so we can query miner status over time with Big Query.

### CODE REVIEW

File: hw_diag/utilities/gcs_shipper.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import requests
+import logging
+from hashlib import sha256
+
+
+log = logging.getLogger()
+log.setLevel(logging.DEBUG)
+
+
+BUCKET_NAME = 'inbound-hotspot-diagnostics'
```
Comment (vpetersson) [inline hw_diag/utilities/gcs_shipper.py:12]: We should probably make this an env variable.

File: bigquery/bq_funcs.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+function transform(line) {
```
Comment (marvinmarnold) [inline bigquery/bq_funcs.js:1]: What's this for?

File: hw_diag/utilities/gcs_shipper.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def generate_hash(public_key):
+    hashable = '%s-%s' % (public_key, datetime.datetime.utcnow().isoformat())
+    return sha256(hashable.encode('utf-8')).hexdigest()
+
+
+def upload_diagnostics(diagnostics):
+    log.info('Submitting diagnostics to GCS bucket.')
+    file_name = generate_hash(diagnostics.get('PK'))
```
Comment (marvinmarnold) [inline hw_diag/utilities/gcs_shipper.py:28]: Could you add a comment explaining why we need to hash. Not immediately clear to me what we are trying to accomplish.

File: bigquery/bq_schema.json [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    "type": "TIMESTAMP"
+  },
+  {
+    "mode": "REQUIRED",
+    "name": "last_updated",
+    "type": "STRING"
+  },
+  {
+    "mode": "NULLABLE",
+    "name": "E0",
```
Comment (marvinmarnold) [inline bigquery/bq_schema.json:15]: Let's finally start using sensible names for these fields now that we are doing greenfield work. Yea, I know annoying, you'll have to add a mapping.

File: bigquery/bq_schema.json [hunk 5]
```diff
@@ -0,0 +1,193 @@
+{ "BigQuery Schema":
+[
+  {
+    "mode": "REQUIRED",
+    "name": "last_updated_ts",
+    "type": "TIMESTAMP"
+  },
+  {
+    "mode": "REQUIRED",
+    "name": "last_updated",
```
Comment (marvinmarnold) [inline bigquery/bq_schema.json:10]: Are you intentionally leaving both last_updated fields in?

File: hw_diag/utilities/gcs_shipper.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    % BUCKET_NAME
+
+
+def generate_hash(public_key):
+    hashable = '%s-%s' % (public_key, datetime.datetime.utcnow().isoformat())
+    return sha256(hashable.encode('utf-8')).hexdigest()
+
+
+def upload_diagnostics(diagnostics):
+    log.info('Submitting diagnostics to GCS bucket.')
```
Comment (marvinmarnold) [inline hw_diag/utilities/gcs_shipper.py:27]: Could you add `BUCKET` to the log message?

**Review by marvinmarnold — DISMISSED** [review]
Great work. Left a few minor comments. It also needs documentation, especially for the GCS stuff.

File: bigquery/bq_funcs.js (hunk 2 above)
Comment (robputt) [reply bigquery/bq_funcs.js:1]: For GCloud Dataflow you need to have a function on the ingestion, it is used to transform the data, in our case it's already in the right format so we just return it.

Comment (marvinmarnold) [reply bigquery/bq_funcs.js:1]: Please include this in the documentation. Also explicitly how to set things up and connect it.

File: bigquery/bq_schema.json (hunk 5 above)
Comment (robputt) [reply bigquery/bq_schema.json:10]: We retain the last_updated string to observe the original data format whilst adding a new field in timestamp form for indexing purposes.

File: bigquery/bq_schema.json (hunk 4 above)
Comment (robputt) [reply bigquery/bq_schema.json:15]: Lets open a separate story for this, the current data format is documented and agreed and has compatibility with all parts of the system, if we wish to rename the keys we should handle separately and refactor all instances.

**Review by marvinmarnold — COMMENTED** [review]
Code looks good. Pending some documentation about what needs to be done in GCS to correctly setup the environment.

File: GCS_DIAGNOSTICS_SHIPPING.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# Google Cloud Storage Diagnostics Shipping
+
+HM-Diag provides the capability to periodically ship hotspot diagnostic
+statistics to Google Cloud Storage for centralised fleet management / 
+feedback for services such as the Nebra Dashboard. Once ingested to Google
+Cloud Storage then DataFlow can be used to marshall the statistics into
+BigQuery for easy consumption with an SQL-like query language.
+
+### Configure Google Cloud Storage Buckets
```
Comment (vpetersson) [inline GCS_DIAGNOSTICS_SHIPPING.md:9]: Need to disable 'Prevent Public Access'

File: hw_diag/utilities/gcs_shipper.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+from hashlib import sha256
+
+
+log = logging.getLogger()
+log.setLevel(logging.DEBUG)
+
+
+BUCKET_NAME = os.getenv(
+    'DIAGNOSTICS_GCS_BUCKET',
+    'inbound-hotspot-diagnostics'
```
Comment (vpetersson) [inline hw_diag/utilities/gcs_shipper.py:15]: Please change this to 'helium-miner-data'

File: GCS_DIAGNOSTICS_SHIPPING.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  anonymous users could read diagnostics data for all hotspots.
+- Create another bucket for storing DataFlow ingestion scratch space, e.g.
+  "hotspot-ingestion-scratch". This bucket does not require any specific
+  ACL / roles. Google Cloud will deal with this for you later when you configure
+  DataFlow.
+- Upload the contents of the bigquery folder of this repository (bq_funcs.js &
+  bq_schema.json) into the DataFlow scratch space bucket. This will be used later
+  by the DataFlow service.
+  
+### Define GCS Bucket in Balena Fleet
```
Comment (vpetersson) [inline GCS_DIAGNOSTICS_SHIPPING.md:38]: This is only needed if we need to override for dev reasons.

File: GCS_DIAGNOSTICS_SHIPPING.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  + Encryption: Google Managed Key
+
+### Configure DataFlow Streaming Job
+- Use the left-hand drawer to navigate to the DataFlow service.
+- Click "Create Job From Template" and configure the job as follows:
+  + Name: hotspot_diagnostics_ingestion_stream
+  + Region: Should be the same region as your BigQuery data set from the previous step
+  + Template: Text Files on Cloud Storage to BigQuery
+- Configure the job parameters as follows:
+  + JavaScript UDF path in Cloud Storage: gs://<your_dataflow_scratch_bucket_name>/bq_funcs.js
```
Comment (vpetersson) [inline GCS_DIAGNOSTICS_SHIPPING.md:59]: helium-miner-data-scratch

File: GCS_DIAGNOSTICS_SHIPPING.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+### Configure BigQuery Data Set
+- Use the left-hand drawer to navigate to the BigQuery service.
+  + or head here https://console.cloud.google.com/bigquery
+- In the BigQuery explorer click the actions menu next to your project's name and
+  select "Add Data Set". Configure your BigQuery data set as follows:
+  + Data Set ID: Something sensible, e.g. "hotspot_diagnostics_data"
+  + Data Location: The preferred region for the data set.
+  + Encryption: Google Managed Key
+
+### Configure DataFlow Streaming Job
```
Comment (vpetersson) [inline GCS_DIAGNOSTICS_SHIPPING.md:52]: Should copy in the job files prior.

File: bigquery/bq_funcs.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
...
+    return line;
```
Comment (vpetersson) [inline bigquery/bq_funcs.js:2]: That's it?

File: hw_diag/tasks.py [hunk 13]
```diff
@@ -92,4 +93,6 @@ def perform_hw_diagnostics():
     with open('diagnostic_data.json', 'w') as f:
         json.dump(diagnostics, f)
 
+    upload_diagnostics(diagnostics)
```
Comment (vpetersson) [inline hw_diag/tasks.py:96]: What's the frequency of this?

**Review by vpetersson — CHANGES_REQUESTED** [review]
Minor changes.

File: GCS_DIAGNOSTICS_SHIPPING.md (hunk 11 above)
Comment (robputt) [reply GCS_DIAGNOSTICS_SHIPPING.md:52]: See line 34

File: hw_diag/tasks.py (hunk 13 above)
Comment (robputt) [reply hw_diag/tasks.py:96]: Depends on how often diagnostics is ran in APScheduler, currently every 5 mins.

File: GCS_DIAGNOSTICS_SHIPPING.md (hunk 7 above)
Comment (robputt) [reply GCS_DIAGNOSTICS_SHIPPING.md:9]: See line 25

**Review by marvinmarnold — APPROVED** [review]

**Pull request merged**